Updated - topologymanager: Add Merge method to Policy

nolancon opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR is an updated version of #81586: 
"In the current design the Topology Manager generates the merged
hints and select the best hint and the policy indicates whether
or not to admit the pod.

This patch adds Merge abstraction to Topology Manager policies which
allow the them to control on how to merge the permutation hints.
It also update the strict policy to admit only if all resources
aligned to same NUMA."

Now that #84721 has been merged, the changes proposed here are the original commits by @adrianchiris, but rebased to accommodate easier reviewing.

Related Issue: #83478 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

nolancon commented:
/cc @adrianchiris @klueska @lmdaly

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -194,6 +184,11 @@ func mergeProvidersHints(policy Policy, numaNodes []int, providersHints []map[st
 }
+
+func (p *bestEffortPolicy) Merge(providersHints []map[string][]TopologyHint) (TopologyHint, lifecycle.PodAdmitResult) {
+	hint := p.merge(providersHints)
```
we can keep the original name : mergeProvidersHints

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -91,11 +85,36 @@ func iterateAllProviderTopologyHints(allProviderHints [][]TopologyHint, callback
+	return TopologyHint{mergedAffinity, preferred}
+}
+
+func (p *bestEffortPolicy) merge(providersHints []map[string][]TopologyHint) TopologyHint {
```
We can keep the original name, mergeProvidersHints.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_restricted.go:
```diff
@@ -52,7 +52,7 @@ func (p *restrictedPolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.PodAd
 
 func (p *restrictedPolicy) Merge(providersHints []map[string][]TopologyHint) (TopologyHint, lifecycle.PodAdmitResult) {
-	hint := mergeProvidersHints(p, p.numaNodes, providersHints)
+	hint := p.merge(providersHints)
```
we can keep the original (mergeProvidersHints) method name

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	return filteredResourcesHints
+}
+
+func (p *singleNumaNodePolicy) merge(providersHints []map[string][]TopologyHint) TopologyHint {
```
for consistency with other policies, this can be renamed to mergeProvierdsHints.

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+					}
+				}
+			}
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
```
I think this append will need a check:  `if len(filtered) != 0` otherwise an empty `filtered` could be appended and we will end up breaking out early here: https://github.com/kubernetes/kubernetes/blob/54e5b088d1ded78d27325f61aa4c607033c0cee2/pkg/kubelet/cm/topologymanager/policy_single_numa_node.go#L176

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
 }
 
+// Get the narrowest hint aligned to one socket
+// assumption: allResourcesHints are not empty, all hints have SocketAffinity != nil,
```
nit: comment - /SocketAffinity/NUMANodeAffinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
 
+// Get the narrowest hint aligned to one socket
+// assumption: allResourcesHints are not empty, all hints have SocketAffinity != nil,
+// all hints have exactly one socket set in SocketAffinity
```
nit: comment
socket/NUMA node/
/SocketAffinity/NUMANodeAffinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+}
+
+func (p *singleNumaNodePolicy) merge(providersHints []map[string][]TopologyHint) TopologyHint {
+	// Set the default hint to return from this function as an any-socket
```
nit: comment - /any-socket/any-NUMA-node/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	var allResourcesHints [][]TopologyHint
+	for _, hints := range providersHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity with any resource, " +
```
nit: comment - /socket affinity/NUMA node affinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity with resource "+
```
nit: comment - /socket affinity/NUMA node affinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible socket affinities for resource '%s'",
```
nit: comment - /socket affinities/NUMA node affinities

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		}
+	}
+	// In case allProviderHints length is zero it means that we have no
+	// preference for socket affinity. In that case update default hint preferred
```
nit: comment - /socket affinity/NUMA node affinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	// preference for socket affinity. In that case update default hint preferred
+	// to true to allow scheduling.
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No preference for socket affinity from all providers")
```
nit: comment - /socket affinity/NUMA node affinity/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	allResourcesHints = p.filterHints(allResourcesHints)
+	// if no hints or there is a resource with empty hints after filtering, then policy cannot be satisfied
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No hints that align to a single socket.")
```
nit: comment - /socket/NUMA node/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	}
+	for _, hints := range allResourcesHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] No hints that align to a single socket for resource.")
```
nit: comment - /socket/NUMA node/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+
+	found, match := p.getHintMatch(allResourcesHints)
+	if found {
+		klog.Infof("[topologymanager] Strict Policy match: %v", match)
```
nit: Log - /Strict Policy/single-numa-node policy/

nolancon commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		klog.Infof("[topologymanager] Strict Policy match: %v", match)
+		return match
+	}
+	klog.Infof("[topologymanager] Strict Policy no match: %v", defaultHint)
```
nit: Log - /Strict Policy/single-numa-node policy/

klueska commented:
Can you please break this PR into small incremental commits , each with a single, logical change in them? The main problem with the original PR was that it was too hard to review because lots of changes were smashed into a single commit.

To help with this, I took the original PR and broke (most of it) it into these logical changes:
https://github.com/kubernetes/kubernetes/pull/84721/commits

Derek was then able to merge them the same day I put the PR up because it was much easier to review: https://github.com/kubernetes/kubernetes/pull/84721#issuecomment-549625351

It would be good to do the same for these remaining changes.

adrianchiris committed:
Refactor policy-best-effort
- Modularize code with mergePermutation method

adrianchiris committed:
Update policy_none removing canAdmitPodResult
Update unit tests for none_policy
Add Name test for policy_restricted

adrianchiris committed:
Make mergeProviderHints policy-specific:
- Remove need to pass policy and numaNodes as arguments
- Remove PolicySingleNUMANode special case check in policy_best_effort
- Add mergeProviderHints base to policy_single_numa_node for upcoming
commit

adrianchiris committed:
Return defaultAffinity from PolicyBestEffort:
Now that PolicySingleNUMANode is not considered here, return
defaultAffinity as was the original case before previous bug fix

adrianchiris committed:
Add new functionality for single-numa-node policy:
Explanation taken from original commit:
- Change the current method of finding the best hint.
  Instead of going over all permutations, sort the hints and find
  the narrowest hint common to all resources.
- Break out early when merging to a preferred hint is not possible

adrianchiris committed:
Update single-numa-node policy unit tests

adrianchiris committed:
Additional unit tests for Topology Manager methods

nolancon commented:
I've broken this PR into smaller commits to accommodate the review process based on the advice of @klueska. 
I've also marked it as WIP - there are some unit tests failing so it will need additional work.

Commit:
Update filterHints:
- Only append valid preferred-true hints to filtered
- Return true if allResourceHints only consist of
nil-affinity/preferred-true hints: {nil true}, update defaultHint
preference accordingly.

Commit:
Update unit tests:
- Update filterHints test to reflect changes in previous commit.
- Some common test cases achieve differing expected results based on
policy due to independent merge strategies. These cases are moved into
individual policy based test functions.

Commit:
Updates to single-numa-node policy:
- Remove getHintMatch method.
- Replace with simplified versions of mergePermutation and
iterateAllProviderTopologyHints methods - as used in best-effort.
- Remove getHintMatch unit tests.

adrianchiris commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -55,49 +54,59 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+// Merge a TopologyHints permutation to a single hint by performing a bitwise-AND
+// of their affinity masks. At this point, all hints have single numa node affinity
+// and preferred-true.
+func (p *singleNumaNodePolicy) mergePermutation(permutation []TopologyHint) TopologyHint {
```
So the direction is to go back to going over all possible permutations like in strict and best effort policy ?

Commit:
Edit hints returned from policies and unit tests:
- Best Effort Policy: Return hint with nil affinity as opposed to
defaultAffinity when provider has no preference for NUMA affinty or no
possible NUMA affinities.
- Single NUMA Node Policy: Remove defaultHint from mergeProvidersHints.
Instead return appropriate TopologyHint where required.
- Update unit tests to reflect changes. Some test cases moved into
individual policy test functions due to differing returned affinties
per policy.

Commit:
Reinstate canAdmitPodResult in policy_none:
This is to keep consistency with the other policies.
This change may be made across all policies in a future PR, but removing it
from the scope of this PR for now.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	}
+	return filteredResourcesHints
+}
+
```
The standard pattern we tend to follow in kubernetes is to return early based on a condition, rather than having the code structured as you do here. This would transform this function
```
func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
        if len(allResourcesHints) == 0 {
                return nil
        }

        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                if len(oneResourceHints) == 0 {
                        continue
                }

                var filtered []TopologyHint
                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 {
                                filtered = append(filtered, hint)
                        }
                }

                filteredResourcesHints = append(filteredResourcesHints, filtered)
        }

        return filteredResourcesHints
}
```

nolancon replied:
Just testing this out now, it will need a couple of minor tweaks to produce the desired output:
``````

func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                var filtered []TopologyHint
                if len(oneResourceHints) == 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                        continue
                }

                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 && hint.Preferred == true {
                                filtered = append(filtered, hint)
                        }
                }

                filteredResourcesHints = append(filteredResourcesHints, filtered)
        }
        return filteredResourcesHints
}

klueska replied:
I guess it depends on what the desired output is.

My original suggestion would have output exactly what the previous iteration of the code would output. Your new version now *always* does a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`), which wouldn't have happened before.

With this new version (assuming these are the desired semantics), the whole check for `if len(oneResourceHints) == 0` can actually be removed (the `filteredResourcesHints = append(filteredResourcesHints, filtered)` at the bottom should suffice for both cases).

klueska replied:
Maybe the true desired semantics arethe following?

```
func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [ []TopologyHint {
        var filteredResourcesHints [][]TopologyHint
        for _, oneResourceHints := range allResourcesHints {
                var filtered []TopologyHint
                for _, hint := range oneResourceHints {
                        if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 && hint.Preferred == true {
                                filtered = append(filtered, hint)
                        }
                }
                if len(filtered) > 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                }
        }
        return filteredResourcesHints
}
```

nolancon replied:
> I guess it depends on what the desired output is.
> 
> My original suggestion would have output exactly what the previous iteration of the code would output. Your new version now _always_ does a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`), which wouldn't have happened before.
> 
> With this new version (assuming these are the desired semantics), the whole check for `if len(oneResourceHints) == 0` can actually be removed (the `filteredResourcesHints = append(filteredResourcesHints, filtered)` at the bottom should suffice for both cases).

I might be missing something here, but the original function does in fact always do a `filteredResourcesHints = append(filteredResourcesHints, filtered)` in all cases (even if `filtered` is `nil`)? 

The idea being that there is then a check for any `nil` hint after `filterHints` is called: https://github.com/kubernetes/kubernetes/blob/078f754a33bb0fcf2a271c2d95ca16a9d9b81c0a/pkg/kubelet/cm/topologymanager/policy_single_numa_node.go#L174

The version I added achieves the same output as the original.

klueska replied:
There's now too many versions of this to keep track of :) Not sure which one you mean when you say original anymore.

If we do want `nil` added, then it's fine to get rid of the check for `if len(filtered) > 0`, but I still think it's better to omit the check for the following because it is unnecessary:
```
                if len(oneResourceHints) == 0 {
                        filteredResourcesHints = append(filteredResourcesHints, filtered)
                        continue
                }
```
If `len(oneResourceHints) == 0` then the loop just below it won't be entered and we will fall down to the `filteredResourcesHints = append(filteredResourcesHints, filtered)` and perform the same operation as whats inside this conditional.

nolancon replied:
Yes, I think we're on the same page now! 😄 I'll tidy this function up.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+				return true
+			}
+			return false
+		})
```
Is there a reason not to use the standard `bitmask.IsNarrowerThan()` operator here?

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+					resource)
+				// return defaultHint which will fail pod admission
+				return defaultHint
+			}
```
This feels like the wrong thing to do, even though it will work in the end because of the false preference. Why can't it instead just return `TopologyHint{}`, which is clearer.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		klog.Infof("[topologymanager] No preference for NUMA affinity from all providers")
+		defaultHint.Preferred = true
+		return defaultHint
+	}
```
It might be good to expand this comment a bit since we are in the `single-numa-node` policy, and it seems counter-intuitive on the surface to ever allow scheduling if we can't align on a single NUMA node. We should make it clear, however, that we need someway of scheduling pods that don't require *any* aligned resources, which is why we have this special-case. The policy itself only ever kicks in, once some resource requires alignment.

That said, I would expect to return `TopologyHint{nil, true}` here, and not the `defaultHint` as defined here.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	// if no hints or there is a resource with empty hints after filtering, then policy cannot be satisfied
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No hints that align to a single NUMA node.")
+		return defaultHint
```
Again, I would return `TopologyHint{}` here, not defaultHint as defined in this code.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	for _, hints := range allResourcesHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] No hints that align to a single NUMA node for resource.")
+			return defaultHint
```
Should return `TopologyHint{}`

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		return defaultHint
+	}
+
+	allResourcesHints = p.filterHints(allResourcesHints)
```
I would add a comment above this, summarizing everything that is going to happen beyond this point. The comment below can be wrapped into this bigger comment.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	}
+	klog.Infof("[topologymanager] single-numa-node no match: %v", defaultHint)
 	return defaultHint
 }
```
I'm not a fan of how this branches and returns in two places. Since we have a helper function, I would expect `getHintMatch()` to just return whatever our result without the need to inspect it.  If no good match is found, it should return `TopologyHint{}`, which is what we want to return anyway in the case of "no match".

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+// Get the narrowest hint aligned to one NUMA node
+// assumption: allResourcesHints are not empty, all hints have NUMAAffinity != nil,
+// all hints have exactly one NUMA node set in NUMANodeAffinity
+func (p *singleNumaNodePolicy) getHintMatch(allResourcesHints [][]TopologyHint) (bool, TopologyHint) {
```
How is this function different than applying `iterateAllProviderTopologyHints()` over a filtered set of hints (i.e. those that only container a single numa node?

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -164,16 +175,25 @@ func (p *singleNumaNodePolicy) mergeProvidersHints(providersHints []map[string][
+			} else if noAffinityPreferredOnly {
+				defaultHint.Preferred = true
+				return defaultHint
+			}
```
I think we should just get rid of the notion of a `defaultHint` in this function altogether. It confuses things, and it is sufficient to just return an appropriate hint at all points without it.

nolancon replied:
Makes sense, I'll make these changes. My guess is the original idea was to use `defaultHint` with the `defaultAffinity` because when the first PR was created `canAdmitPodResult` used to do a check on `hint.NUMANodeAffinity.Count() != 1`. Whereas now it just checks `!hint.Preferred`, so the affinity is kind of irrelevant.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -100,23 +100,34 @@ func (p *singleNumaNodePolicy) getHintMatch(allResourcesHints [][]TopologyHint)
+	if noAffinityPreferredHints == totalHints {
+		return filteredResourcesHints, true
+	}
+	return filteredResourcesHints, false
```
As discussed offlline, I think there should be better way to handle this.

nolancon replied:
Agreed, but it does feel like something that should be handled in a `filterHints` function :) - Probably need to have another think about this.

nolancon replied:
This bug fix is now removed and can be dealt with in a future commit.
One option is to do a check just before `filterHints` is called to see if `allResourceHints` consists **only** of `{nil, true}` hints. If so, return `{nil, true}`:
``````
duplicate := make([][]TopologyHint, len(allResourcesHints))
for i := range allResourcesHints {
	duplicate[i] = make([]TopologyHint, len(allResourcesHints[i]))
        for j := range duplicate[i] {
        	duplicate[i][j] = TopologyHint{nil, true}
        }
}

if reflect.DeepEqual(duplicate, allResourcesHints) {
        return TopologyHint{nil, true}
}
``````

klueska replied:
Is there a test case that fails because we omit this bug fix?

nolancon replied:
Yes, as follows:
```
Single TopologyHint with Preferred as true and NUMANodeAffinity as nil: 
Expected Topology Hint to be {<nil> true}, got {<nil> false}:
```
The problem being that `filterHints` treats hints `{<nil> true}` and `{<nil> false}` equally by appending a nil `[]TopologyHint` in both cases. 

So in both cases they end up returning  `{<nil> false}` when checked after `filterHints` is called. A bit of a corner case but probably one we still need to catch?

klueska reviewed (changes requested):
This is a first pass of review. I have some more comments about general structure of thing,s but let's try and address what's here first and then come back to these.

nolancon commented on pkg/kubelet/cm/topologymanager/policy_none.go:
```diff
@@ -36,12 +36,6 @@ func (p *nonePolicy) Name() string {
-}
-
 func (p *nonePolicy) Merge(providersHints []map[string][]TopologyHint) (TopologyHint, lifecycle.PodAdmitResult) {
-	return TopologyHint{}, p.canAdmitPodResult(nil)
```
I agree with the 1st comment, probably best to keep symmetry with the other policies and leave it 'as is'. If we want to change all three in future we can revisit it.

nolancon commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -72,7 +72,7 @@ func (p *bestEffortPolicy) Merge(providersHints []map[string][]TopologyHint) (To
 //                 }
 //                 callback(permutation)
-func iterateAllProviderTopologyHints(allProviderHints [][]TopologyHint, callback func([]TopologyHint)) {
+func (p *bestEffortPolicy) iterateAllProviderTopologyHints(allProviderHints [][]TopologyHint, callback func([]TopologyHint)) {
```
Sure - so leave this function generic. If we decide to go the route of using this function in `policy_single_numa_node` is it OK to leave it in `policy_best_effort` to be called?

nolancon replied:
This function is now made generic and moved into `policy.go` as agreed offline.

nolancon commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -105,21 +124,21 @@ func mergeProvidersHints(policy Policy, numaNodes []int, providersHints []map[st
 		if len(hints) == 0 {
 			klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with any resource")
-			allProviderHints = append(allProviderHints, []TopologyHint{{nil, true}})
+			allProviderHints = append(allProviderHints, []TopologyHint{{defaultAffinity, true}})
```
Yes, makes sense - I will update this.

Commit:
Make iterateAllProviderTopologyHints generic:
- Remove policy parameters to make this function generic.
- Move function out of individual policies and into policy.go

Commit:
Refactor filterHints:
- Restructure function
- Remove bug fix for catching {nil true} - To be fixed in later commit
- Restore unit tests to original state for testing filterHints

Commit:
Make mergePermutation generic:
- Remove policy parameters to make function generic
- Move function into top level policy.go

Commit:
Restore original policy none test cases:
Mistakenly overwritten in earlier commit

klueska commented on pkg/kubelet/cm/topologymanager/policy_best_effort.go:
```diff
@@ -48,54 +48,46 @@ func (p *bestEffortPolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.PodAd
+// Merge a TopologyHints permutation to a single hint by performing a bitwise-AND
+// of their affinity masks. The hint shall be preferred if all hits in the permutation
+// are preferred.
+func (p *bestEffortPolicy) mergePermutation(permutation []TopologyHint) TopologyHint {
```
I think this function can become top level as well (i.e. .not per-policy). Filtering in the single-numa node case will ensure we don't hit cases in this function where there is a hint with `NUMANodeAffinity == nil`.

nolancon replied:
Will make this function generic and move into `policy.go`

klueska commented on pkg/kubelet/cm/topologymanager/policy_none_test.go:
```diff
@@ -40,28 +41,44 @@ func TestName(t *testing.T) {
+				},
+			},
+			expectedHint:  TopologyHint{},
+			expectedAdmit: lifecycle.PodAdmitResult{Admit: true},
```
This seems to be a completely different set of tests now for this test case. Is there some way to keep the old tests and then add these as new ones?

nolancon replied:
Looks like there was a mix-up here... `TestNoneCanAdmitPodResult` should have been left as is, and `TestPolicyMerge` should contain these new tests - will fix.

klueska commented on pkg/kubelet/cm/topologymanager/policy_none_test.go:
```diff
@@ -40,28 +41,44 @@ func TestName(t *testing.T) {
+				},
+			},
+			expectedHint:  TopologyHint{},
+			expectedAdmit: lifecycle.PodAdmitResult{Admit: true},
```
Same comment as above.

nolancon replied:
Responded above.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+	// affinity with an unpreferred allocation. This will only be returned if
+	// no better hint can be found when merging hints from each hint provider.
+	defaultAffinity, _ := bitmask.NewBitMask(p.numaNodes...)
+	bestHint := TopologyHint{defaultAffinity, false}
```
For this policy, it should suffice to just initialize `bestHint` to `TopologyHint{}` (because we know we will always fail pod admission regardless of the Affinity mask, so long as preferred is false.

nolancon replied:
If we initialize to `TopologyHint{}`, we will also need to remove the check mentioned in the comment below - as checking `NUMANodeAffinty` of a `nil` `bestHint` will error out.

klueska replied:
It just seems odd to initialize it to something more complicated than its default type if we don't actually care about it having this value. I'd rather have the non-null check below to be more explicit.

nolancon replied:
Revised checks in response to below comment.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		// do not update defaultHint
+		if mergedHint.NUMANodeAffinity.Count() == bestHint.NUMANodeAffinity.Count() {
+			return
+		}
```
This check seems unnecessary for this policy (given the check right above this). However, I think we need to add a chack to make sure that `bestHint` is always updated to the narrower of the two hints.

nolancon replied:
Without this check, a `bestHint.NUMANodeAffinity` of `{...001}` will be replaced by a `mergedHint.NUMANodeAffinity` of `{...0010}` or `{...0100}` etc as they occur.

As for a check to make sure that `bestHint` is always updated to the narrower of the two hints... `filterHints` ensures that only hints with `NUMANodeAffinity.Count() == 1` are kept... so we should only ever end up with `bestHint` that has `NUMANodeAffinity` of width 0 or 1 - which is handled in the first check.

klueska replied:
`isNarrowerThan()` handles this case. It chooses the one with the lowest order bits to be considered narrower.
 https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/bitmask/bitmask.go#L129

nolancon replied:
Sure, so our three checks could be as follows: (including check for nil `bestHint`)
````
// Only consider mergedHints that result in a NUMANodeAffinity == 1 to
// replace the current bestHint.
if mergedHint.NUMANodeAffinity.Count() != 1 {
       return
}

// If the current bestHint is nil, replace with mergedHint.
if bestHint == (TopologyHint{}) {
       bestHint = mergedHint
       return
}

// Only consider mergedHints that have a narrower NUMANodeAffinity
// than the NUMANodeAffinity in the current bestHint.
if !mergedHint.NUMANodeAffinity.IsNarrowerThan(bestHint.NUMANodeAffinity) {
     return
}

klueska replied:
Looks good. Only slight modification I'd make is to check `if bestHint.NUMANodeAffinity == nil` instead of `bestHint == (TopologyHint{})` since it's the fact that that field is `nil` that we have to protect against.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+		filteredResourcesHints = append(filteredResourcesHints, filtered)
+	}
+	return filteredResourcesHints
+}
```
Please see https://github.com/kubernetes/kubernetes/pull/85798#discussion_r359801185 for my latest comments on this function.

nolancon replied:
Replied under this comment also.

klueska commented on pkg/kubelet/cm/topologymanager/policy_single_numa_node.go:
```diff
@@ -52,8 +54,122 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+			klog.Infof("[topologymanager] TopologyHints for resource '%v': %v", resource, hints[resource])
+			allResourcesHints = append(allResourcesHints, hints[resource])
+		}
+	}
```
This all feels very generic (i.e reusable across all policies). Does it make sense to pull this into a top level function as well, something like `gatherAllProviderHints()`?

nolancon replied:
I agree, but I think there are some subtle differences between policies and already a lot of change so I'm inclined to leave it 'as is' for now and maybe revisit later?

klueska replied:
Fair enough. Sounds good to me.

Commit:
Update checks in mergeProvidersHints:
- Initialize best Hint to TopologyHint{}
- Update checks.
- Move generic unit test case into policy specific tests and updated
expected outcome to reflect changes.

Commit:
Restore policy_test.go to upstream
Following commits will contain incremental changes to this file to ease
review process and ensure all tests are accounted for.

Commit:
Move test case 'TopologyHint not set' into individual policy tests

Commit:
Move test case 'HintProvider returns empty non-nil map[string][]TopologyHint' into specific policy tests.

Commit:
Move test case "HintProvider returns -nil map[string][]TopologyHint from provider" into specific policy tests

Commit:
Move test case "HintProvider returns empty non-nil map[string][]TopologyHint from
provider" into specific policy tests.

Commit:
Move test case "Single TopologyHint with Preferred as true and
NUMANodeAffinity as nil" into specific policy tests.

Commit:
Move test case "Single TopologyHint with Preferred as false and
NUMANodeAffinity as nil" into specific policy tests.

Commit:
Move test case "Two providers, 1 hint each, no common mask" into
specific policy tests.

Commit:
Move test case "Two providers, 1 hint each, same mask, 1 preferred, 1
not 1/2" into specific policy test.

Commit:
Move test case "Two providers, 1 hint each, same mask, 1 preferred, 1
not 2/2" into specific policy tests

Commit:
Move test case "Two providers, 1 with 2 hints, 1 with single
non-preferred hint matching" into specific policy tests

Commit:
Update "Single NUMA hint generation" expected affinity to nil

Commit:
Use reflect.DeepEqual check in policy_test.go

klueska commented:
Things are coming together nicely now. Few remaining comments:
1) We need to work out a solution for https://github.com/kubernetes/kubernetes/pull/85798#discussion_r358746395
2) Is there a way to clean up the diff in `policy_test.go` it's hard to verify that no tests were removed and only new tests were added.

nolancon commented:
> Things are coming together nicely now. Few remaining comments:
> 
> 1. We need to work out a solution for [#85798 (comment)](https://github.com/kubernetes/kubernetes/pull/85798#discussion_r358746395)
> 2. Is there a way to clean up the diff in `policy_test.go` it's hard to verify that no tests were removed and only new tests were added.

Difficult to keep the diff tidy with so much duplication in the file, I can break up the changes to `policy_test.go` into smaller commits to make it clearer so we're not actually losing any test cases?

nolancon commented:
> > Things are coming together nicely now. Few remaining comments:
> > 
> > 1. We need to work out a solution for [#85798 (comment)](https://github.com/kubernetes/kubernetes/pull/85798#discussion_r358746395)
> > 2. Is there a way to clean up the diff in `policy_test.go` it's hard to verify that no tests were removed and only new tests were added.
> 
> Difficult to keep the diff tidy with so much duplication in the file, I can break up the changes to `policy_test.go` into smaller commits to make it clearer so we're not actually losing any test cases?

I've broken up the `policy_test.go` changes into commits of each individual test change.
This commit https://github.com/kubernetes/kubernetes/pull/85798/commits/5c9ad04b747b231bed2bac3142ec80575639cbe7 restores the file to current upstream version. All commits **after** this are the individual changes, so reviewing by commit is easiest. I can squash them all back into one commit when review is complete if desired.

klueska committed:
Cleanup use of defaultAffinity in mergePermutation of TopologyManager

klueska committed:
Rename TopologyManager test TestPolicyBestEffortMerge for consistency

klueska committed:
Update TopologyManager single-numa-node logic to handle "don't cares"

The logic has been updated to match the logic of the best-effort policy
except in two places:

1) The hint filtering frunction has been updated to allow "don't care"
hints encoded with a `nil` affinity mask, to pass through the filter in
addition to hints that have just a single NUMA bit set.

2) After calculating the `bestHint` we transform "don't care" affinities
encoded as having all NUMA bits set in their affinity masks into "don't
care" affinities encoded as `nil`.

klueska commented:
/approve
/lgtm

I'm a big fan of being able to walk commit by commit through a PR and see all the final changes build up one-by one (instead of having a change in an earlier commit be overwritten by a later one). A good example of this is the breakdown for the moving of the tests in `policy_test.go`.  It would be nice if the other commits built on each other similarly, but we've iterated on this one quite alot, and I'm happy with the final result, so let's go ahead and merge this one. In the future, let's see if we can break them down as described all the way through.

nolancon commented:
/kind cleanup
